### ISSUE TITLE
OCDS: Flattening record packages

### ISSUE THREAD

**Issue by duncandewhurst** [OpenDataServices/cove#1152]
Some quick notes from a discussion with @kindly  on this today:

The primary use case we have in is publishers sharing draft data with the helpdesk in record package format and helpdesk analysts wanting to flatten this to help give feedback (e.g. it's easier to check all the values of a given field by looking at a spreadsheet than by reviewing a JSON file).

Records have the following components:
 * Compiled release - easiest to flatten, using existing flatten-tool functionality. could include in minimal version of record package flattening
 * Releases list, which can be either linked releases or flattened releases. Key issue is the field is `oneOf` and it isn't possible to definitely tell which has been provided. Also the list could be mixed. Would require more work.
  * Linked releases - assume we won't fetch full releases for flattening, Q: would flattening the list of linked releases (url, date and tag) be useful?
  * Embedded releases - would need to combine the release lists from all records into one list - not sure where this functionality would sit between CoVE and flatten-tool. Could require lots of work.
 * Versioned release - we don't have approach to flattening this and it is very rarely used by published, would leave out of functionality for now

We also discussed whether compiled and embedded releases should be flattened into the same spreadsheet.

* Flattening to separate spreadsheets would be easier to implement and it would be easier to use for users who just wanted to work with one or the other (no filtering out would be required). 
* Flattening to one spreadsheet would be harder to implement but it would make it easier for users who wanted to do analysis across compiled and individual releases (we don't have a specific use case for that in mind).

We also discussed whether this would sit better in CoVE/flatten-tool or OCDSKit/-web (possibly via the tabulate command supporting a spreadsheet output).

Seeking feedback from @yolile @romifz @jpmckinney @mrshll1001@pindec and others on:

* How important is this for the helpdesk
* Other use cases for flattening record packages
* Views on which elements of records are important to flatten
* Flattening records into one or multiple spreadsheets
* Where this functionality belongs

**Comment by jpmckinney** [comment]
If flatten tool accepted standard input (which I had tried to get merged in https://github.com/OpenDataServices/flatten-tool/pull/246), then, to get all releases and all compiled releases into one spreadsheet, couldn't we just do:

`jq '.records[].releases + [.records[].compiledRelease]' fictional-example/1.1/record/ocds-213czf-000-00001-withversions.json | flatten-tool --root-id ocid --root-is-list --main-sheet-name releases`

(As flatten-tool doesn't accept standard input, you can still do it now using an intermediate file.)

To get only the releases or only the compiled releases, you can just change the jq filter to one side of the `+` addition. If you want to exclude linked releases, you can use the filter `'.records[].releases | select(.[].ocid)'`

I don't think we should do any option that is a lot of work, considering this already works and allows the analyst to decide whether to have one or multiple spreadsheets…

Note that I tried to flatten a versioned release and I got `Cannot convert [OrderedDict([('releaseID', 'ocds-213czf-000-00001-04-award'), ('releaseDate', '2010-05-10T09:30:00Z'), ('releaseTag', ['award']), ('value', '45233130.0')])] to Excel`.

**Comment by duncandewhurst** [comment]
The need to flatten the compiled releases from a record package came up in the data use training yesterday.

It's easy enough to do this using `jq` but those types of command line tools are only suited to more technical users.

For users with a skills more suited to web tools (Data Review Tool and ocdskit-web) and spreadsheets it would be good to support flattening of record packages, with a spreadsheet of the compiled releases being the most useful element (since compiled releases are much easier to analyse than individual releases and most questions can be answered using them).

**Comment by jpmckinney** [comment]
If implemented, I think this should go into ocdskit-web, since it's constructed as a set of small, discrete operations that users can perform in the browser, as an alternative to the command line. If agreed, we can close this issue, and re-open one there.

**Comment by robredpath** [comment]
I think there's a few things to consider here:
- It might be appropriate for this functionality to exist in multiple places for different users. That doesn't necessarily mean we need to abstract it out to a library, but we shouldn't think that this can only be in one place
- ocdskit-web is probably the right place for it to live for analysis use 
- CoVE is something of a one-stop-shop for publisher feedback on their data, so I can imagine it living there - instead of asking people to go to another website. 
- If there's a large file, uploading it can be very time-consuming and potentially fragile if there's any interruptions to the upload. If this does go into CoVE, then there's already a function to download from a URL. If this goes into ocdskit-web, we should look at implementing something similar. 
- It might be that there's a relatively simple new product here, whereby a simple web interface for cmdline tools (ie, ocdskit-web) is combined with a private workspace and a high spec machine (eg GCP, AWS) per user, with some management to build, start, stop and destroy the machine on-demand. A little bit of dev and some $$$ to a cloud provider might bridge this gap. 

I've reviewed the PR to get flatten-tool to accept STDIN and I think it's just stalled. I've added it to Trello for us to consider.

**Comment by jpmckinney** [comment]
Good points. If we opt for multiple interfaces for the same operation, we should either use a common library for the implementation, or have one interface use another's API to calculate a response (e.g. CoVE can send a URL to ocdskit-web to calculate the flattened records).

**Comment by duncandewhurst** [comment]
I agree this makes sense for ocdskit-web, but I think it should also be available in the data review tool.

One of the original reasons for having the flattening functionality available in the data review tool was to make it easier for less technical users (e.g. procurement officials tasked with OCDS implementation, OCP programme managers etc.) to troubleshoot the errors reported by the data review tool (e.g. it is much easier for those users to spot the incorrect date format by looking down a column in a spreadsheet, than it is by trying to review a JSON file).

I think that use case applies to flattening records too.

As Rob says, the data review tool has been and still is a one-stop shop and it doesn't make sense to say 'you can flatten releases here but to flatten records you need to go elsewhere'.

> If there's a large file, uploading it can be very time-consuming and potentially fragile if there's any interruptions to the upload. If this does go into CoVE, then there's already a function to download from a URL. If this goes into ocdskit-web, we should look at implementing something similar.

+1

**Comment by jpmckinney** [comment]
As in my earlier comment, I'm open to multiple interfaces for the same operation, with the only caveat that we not duplicate the effort of implementing the software in each interface.
